crates/sources/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -0,0 +1,57 @@
+use super::graph_api::{RegisteredPools, Token, UniV3SubgraphClient};
+use crate::recent_block_cache::Block;
+use crate::token_pair::TokenPair;
+use anyhow::Result;
+use ethcontract::{H160, U256};
+use num::BigInt;
+use reqwest::Client;
+use std::collections::{HashMap, HashSet};
+
+#[async_trait::async_trait]
+pub trait PoolFetching: Send + Sync {
+    async fn fetch(&self, token_pairs: HashSet<TokenPair>, at_block: Block) -> Result<Vec<Pool>>;
```

Comment:
I think expecting a `HashSet<TokenPair>` is nice for everybody implementing `PoolFetching` because it automatically means you don't have to think about duplicates but I expect this API to be quite unwieldy for the users.
Also it can a be a bit of work associating back the unique `Pool`s of the result with the duplicated `TokenPair` originals. This is probably not needed for fetching liquidity but  good to keep in mind.

I think one learning we had from the `PriceEstimating` trait is that it's nice for the people using the API if it's stream based. This makes it way easier to deal with long running requests because you can abort them any time and still keep the results you gathered until that point.

Both of these suggestions make it more difficult to implement the trait but I think ease of use usually trumps ease of implementation.

To conclude my suggestion would be something like this:
```rust
async fn fetch(&self, token_pairs: impl IntoIterator<Item = TokenPair>, at_block: Block) -> BoxStream<Result<Pool>>
```

Reply:
There a couple of considerations that depend on the possible implementations.
1. Is there shared work when fetching multiple pools (like batch calls)? If not, then the trait can be simpler by only taking one token pair.
2. Is it possible that returned pools become ready at different points in time and that callers benefit from receiving them early? If yes then we can return Stream.
3. What is the correspondence of the input token pairs and output vectors? Is it one to one so they're always the same length or can one input map to 0 or more than 1 output?
4. Should the input be taken by value or reference? We could instead take `&HashSet` or `&[TokenPair]`. This can be more efficient if the trait implementations don't need ownership of the data and callers want to reuse it. On the other hand it makes the trait harder to use because of extra lifetimes. This happens in the PriceEstimating trait too.

I'm guessing that there is only going to be one implementation of this trait so I'm fine with keeping the trait as more of a placeholder for now and changing as we see what that implementation needs.

Note that you probably can't use `impl Iterator` in a trait function parameter because the trait wouldn't be object safe anymore. You could use `Box<dyn Iterator>` or `&mut dyn Iterator`.